metadata-ingestion/src/datahub/ingestion/source/ldap.py

```diff
@@ -389,21 +437,27 @@ def build_corp_group_mce(self, attrs: dict) -> Optional[MetadataChangeEvent]:
                 if self.config.group_attrs_map["displayName"] in attrs
                 else None
             )
-            return MetadataChangeEvent(
-                proposedSnapshot=CorpGroupSnapshotClass(
-                    urn=f"urn:li:corpGroup:{full_name}",
-                    aspects=[
-                        CorpGroupInfoClass(
-                            email=email,
-                            admins=admins,
-                            members=members,
-                            groups=[],
-                            description=description,
-                            displayName=displayName,
-                        )
-                    ],
-                )
+            group_snapshot = CorpGroupSnapshotClass(
+                urn=f"urn:li:corpGroup:{full_name}",
+                aspects=[
+                    CorpGroupInfoClass(
+                        email=email,
+                        admins=admins,
+                        members=members,
+                        groups=[],
+                        description=description,
+                        displayName=displayName,
+                    ),
+                    StatusClass(removed=False),
+                ],
             )
+
+            if full_name:
+                self.stale_entity_removal_handler.add_entity_to_state(
+                    type="corpGroup", urn=builder.make_group_urn(full_name)
+                )
+
+            return MetadataChangeEvent(proposedSnapshot=group_snapshot)
```

Comment:
you can simplify this file even more! instead of manually calling `add_entity_to_state` and creating the status aspects, you can use the helpers that we have. 

I know our docs aren't amazing at explaining these helpers, so apologies for that.

Example:

https://github.com/datahub-project/datahub/blob/308b4eae87b20fa1a557e78ac0cc0329f63e6e4e/metadata-ingestion/src/datahub/ingestion/source/tableau.py#L1612-L1616

Reply:
Yes, np, I will check and make that update. I have another question related to def close(self):
    def close(self) -> None:
```
        """Closes the Source."""
        self.ldap_client.unbind()
        self.prepare_for_commit(). <---- not needed anymore
        super().close()
```
I believe with a recent update you made at the parent class I don't need    self.prepare_for_commit() anymore (it was required before)

Reply:
Yup that's right - the super().close() call internally calls prepare_for_commit, so you don't have to do it again

Reply:
> you can simplify this file even more! instead of manually calling `add_entity_to_state` and creating the status aspects, you can use the helpers that we have.
> 
> I know our docs aren't amazing at explaining these helpers, so apologies for that.
> 
> Example:
> 
> https://github.com/datahub-project/datahub/blob/308b4eae87b20fa1a557e78ac0cc0329f63e6e4e/metadata-ingestion/src/datahub/ingestion/source/tableau.py#L1612-L1616

I still need to complete this last request for the change

Reply:
I am not bale to find these utilities (auto_status_aspect) at our corporate Datahub version 9.0.0. It is probably was added recently, but because we don't have it, I cannot even test this now.

Reply:
Got it - yep it was merged Oct 13 and released as part of 0.9.1 a few weeks later

Note that it is possible to use newer cli versions with older DataHub releases, so you should be able to update to 0.9.1

Reply:
I think I understand how to auto_status_aspect() works. Essentially, I pass in "work units" and I got back an updated "workunits" where for all entities  StatusClass(removed=False).

 So, I can remove 

 StatusClass(removed=False),

and use the following to add  StatusClass:

yield from auto_status_aspect(self.stale_entity_removal_handler.gen_removed_entity_workunits())

This part I have tested and updated. However, I am not sure about "add_entity_to_state" functionality. Could you please elaborate on this slight more?

Reply:
Sure thing. The eventual code will should look like this:

https://github.com/datahub-project/datahub/blob/44cfd21a653e23f869526a3c1a83e11d51b42a86/metadata-ingestion/src/datahub/ingestion/source/tableau.py#L1616-L1619

Your understanding of `auto_status_aspect` is almost right. The one difference - you'd need to rename the existing `get_workunits` to `get_workunits_internal`, and create a new get_workunits method with the contents above that calls the get_workunits_internal one. This way, auto_status_aspect applies on everything produced by the rest of the source.

`auto_stale_entity_removal` is a similar automatic function - it calls add_entity_to_state for each entity, and then calls gen_removed_entity_workunits once everything has been produced. That way, you can remove all calls to add_entity_to_state and gen_removed_entity_workunits in the rest of your code.